[Temporal] Add PlainYearMonth type

catamorphism opened the pull request:
#### 1644afaa9af7533403b79fdc8cbf6949dfa455a0
<pre>
[Temporal] Add PlainYearMonth type
<a href="https://bugs.webkit.org/show_bug.cgi?id=303245">https://bugs.webkit.org/show_bug.cgi?id=303245</a>

Reviewed by Yusuke Suzuki.

Implement basic support for PlainYearMonth, without its methods.

Test: JSTests/stress/temporal-plainyearmonth.js
* JSTests/stress/temporal-plainyearmonth.js: Added.
(shouldBe):
(shouldThrow):
(const.yearMonth.new.Temporal.PlainYearMonth):
* JSTests/test262/config.yaml:
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/DerivedSources.make:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/runtime/ISO8601.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::visitChildrenImpl):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::plainYearMonthStructure):
* Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
(JSC::TemporalCalendar::isoDateFromFields):
(JSC::TemporalCalendar::isoDateAdd):
(JSC::TemporalCalendar::calendarDateUntil):
* Source/JavaScriptCore/runtime/TemporalObject.cpp:
(JSC::createPlainYearMonthConstructor):
* Source/JavaScriptCore/runtime/TemporalPlainYearMonth.cpp: Added.
(JSC::TemporalPlainYearMonth::create):
(JSC::TemporalPlainYearMonth::createStructure):
(JSC::TemporalPlainYearMonth::TemporalPlainYearMonth):
(JSC::TemporalPlainYearMonth::finishCreation):
(JSC::TemporalPlainYearMonth::visitChildrenImpl):
(JSC::TemporalPlainYearMonth::tryCreateIfValid):
(JSC::TemporalPlainYearMonth::monthCode const):
* Source/JavaScriptCore/runtime/TemporalPlainYearMonth.h: Added.
* Source/JavaScriptCore/runtime/TemporalPlainYearMonthConstructor.cpp: Added.
(JSC::TemporalPlainYearMonthConstructor::create):
(JSC::TemporalPlainYearMonthConstructor::createStructure):
(JSC::TemporalPlainYearMonthConstructor::TemporalPlainYearMonthConstructor):
(JSC::TemporalPlainYearMonthConstructor::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/TemporalPlainYearMonthConstructor.h: Added.
* Source/JavaScriptCore/runtime/TemporalPlainYearMonthPrototype.cpp: Added.
(JSC::TemporalPlainYearMonthPrototype::create):
(JSC::TemporalPlainYearMonthPrototype::createStructure):
(JSC::TemporalPlainYearMonthPrototype::TemporalPlainYearMonthPrototype):
(JSC::TemporalPlainYearMonthPrototype::finishCreation):
(JSC::JSC_DEFINE_CUSTOM_GETTER):
* Source/JavaScriptCore/runtime/TemporalPlainYearMonthPrototype.h: Added.

Canonical link: <a href="https://commits.webkit.org/303780@main">https://commits.webkit.org/303780@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/76f86858e2571fd534cf23248fd5a393f79a417f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/133565 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/6070 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/44729 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/141125 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/85617 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/6590 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/5934 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/102186 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/85617 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/136512 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/162/builds/4674 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/119756 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/82982 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/4552 "Passed tests") | [  ~~🧪 api-mac-debug~~](https://ews-build.webkit.org/#/builders/165/builds/2161 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [⏳ 🛠 wpe-cairo ](https://ews-build.webkit.org/#/builders/WPE-Cairo-Build-EWS "Waiting in queue, processing has not started yet") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/125642 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/113689 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/37869 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/143774 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/132079 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/5742 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/38450 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/110562 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/5822 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/4928 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/110744 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28079 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/4410 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/116008 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/59491 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/5794 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/34322 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/165042 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/5641 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/69252 "Built successfully") | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/43130 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/163/builds/5885 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/5749 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/runtime/ISO8601.h:
```diff
@@ -253,6 +253,18 @@ constexpr bool isYearWithinLimits(double year)
 }
 
+// https://tc39.es/proposal-temporal/#sec-temporal-isoyearmonthwithinlimits
+constexpr bool isYearMonthWithinLimits(double year, double month)
```
Why is it taking double?

catamorphism replied:
Fixed in [2c8b496](https://github.com/WebKit/WebKit/pull/54564/commits/2c8b4967a7e40e5ff79b9d260fbbfdd76548489c). I also added an overload of `isYearWithinLimits` to make sure nothing is getting silently cast to double. The `double` overload of `isYearWithinLimits` needs to be there because that's what's used to convert years provided in JS objects from double to int32.

Constellation commented on Source/JavaScriptCore/runtime/ISO8601.h:
This is not a change in this patch, but whenever taking double, that should be wrong.

catamorphism committed:
[Temporal] Add PlainYearMonth type
https://bugs.webkit.org/show_bug.cgi?id=303245

Reviewed by Yusuke Suzuki.

Implement basic support for PlainYearMonth, without its methods.

Test: JSTests/stress/temporal-plainyearmonth.js
* JSTests/stress/temporal-plainyearmonth.js: Added.
(shouldBe):
(shouldThrow):
(const.yearMonth.new.Temporal.PlainYearMonth):
* JSTests/test262/config.yaml:
* JSTests/test262/expectations.yaml:
* Source/JavaScriptCore/CMakeLists.txt:
* Source/JavaScriptCore/DerivedSources.make:
* Source/JavaScriptCore/Sources.txt:
* Source/JavaScriptCore/runtime/ISO8601.h:
* Source/JavaScriptCore/runtime/JSGlobalObject.cpp:
(JSC::JSGlobalObject::init):
(JSC::JSGlobalObject::visitChildrenImpl):
* Source/JavaScriptCore/runtime/JSGlobalObject.h:
(JSC::JSGlobalObject::plainYearMonthStructure):
* Source/JavaScriptCore/runtime/TemporalCalendar.cpp:
(JSC::TemporalCalendar::isoDateFromFields):
(JSC::TemporalCalendar::isoDateAdd):
(JSC::TemporalCalendar::calendarDateUntil):
* Source/JavaScriptCore/runtime/TemporalObject.cpp:
(JSC::createPlainYearMonthConstructor):
* Source/JavaScriptCore/runtime/TemporalPlainYearMonth.cpp: Added.
(JSC::TemporalPlainYearMonth::create):
(JSC::TemporalPlainYearMonth::createStructure):
(JSC::TemporalPlainYearMonth::TemporalPlainYearMonth):
(JSC::TemporalPlainYearMonth::finishCreation):
(JSC::TemporalPlainYearMonth::visitChildrenImpl):
(JSC::TemporalPlainYearMonth::tryCreateIfValid):
(JSC::TemporalPlainYearMonth::monthCode const):
* Source/JavaScriptCore/runtime/TemporalPlainYearMonth.h: Added.
* Source/JavaScriptCore/runtime/TemporalPlainYearMonthConstructor.cpp: Added.
(JSC::TemporalPlainYearMonthConstructor::create):
(JSC::TemporalPlainYearMonthConstructor::createStructure):
(JSC::TemporalPlainYearMonthConstructor::TemporalPlainYearMonthConstructor):
(JSC::TemporalPlainYearMonthConstructor::finishCreation):
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/JavaScriptCore/runtime/TemporalPlainYearMonthConstructor.h: Added.
* Source/JavaScriptCore/runtime/TemporalPlainYearMonthPrototype.cpp: Added.
(JSC::TemporalPlainYearMonthPrototype::create):
(JSC::TemporalPlainYearMonthPrototype::createStructure):
(JSC::TemporalPlainYearMonthPrototype::TemporalPlainYearMonthPrototype):
(JSC::TemporalPlainYearMonthPrototype::finishCreation):
(JSC::JSC_DEFINE_CUSTOM_GETTER):
* Source/JavaScriptCore/runtime/TemporalPlainYearMonthPrototype.h: Added.

Canonical link: https://commits.webkit.org/303780@main
